kiwi_lint/function_based_views.py

```diff
@@ -0,0 +1,130 @@
+        django.setup()
+
+        self.__installed_apps = self._get_django_project_apps()
+        print(self.__installed_apps)
+
+        project_urls = import_module(main_urls_package)
+
+        self.url_mapping = self._get_url_view_mapping(project_urls.urlpatterns)
+        self.url_mapping = self._prune_url_mapping(self.url_mapping, self.__installed_apps)
+        self.view_files = set(module for module, _ in self.url_mapping.values())
+
+    def _get_django_project_apps(self):
+        installed_apps = deepcopy(settings.INSTALLED_APPS)  # do not mutate the original
+
+        for _filter in self.installed_apps_function_filters:
+            # allow void filters that just mutate the list
+            res = list(_filter(installed_apps) or [])
+            installed_apps = res if res else installed_apps
+
+        return installed_apps
+
+    @staticmethod
+    def _get_url_view_mapping(urlpatterns):
+        def helper(urlpatterns, prefix='^', acc=None):
+            if acc is None:
+                acc = OrderedDict()
+
+            for url in urlpatterns:
+                if isinstance(url, URLPattern):
+                    key = prefix + url.pattern.regex.pattern.strip('^')
+                    acc[key] = (url.callback.__module__, url.callback.__name__)
+
+                elif isinstance(url, URLResolver):
+                    helper(url.url_patterns, prefix + url.pattern.regex.pattern.strip('^$'), acc)
+            return acc
+
+        return helper(urlpatterns)
+
+    @staticmethod
+    def _prune_url_mapping(url_mapping, installed_apps):
+        pruned = url_mapping.__class__()
+        for k, (module, name) in url_mapping.items():
+            if any(module.startswith(app) for app in installed_apps):
+                pruned[k] = (module, name)
+
+        return pruned
+
+    def visit_module(self, module):
+        if module.name in self.view_files:
+            if hasattr(self, 'visit_views_module'):
+                self.visit_views_module(module)
+
+    def leave_module(self, module):
+        if module.name in self.view_files:
+            if hasattr(self, 'visit_views_module'):
+                self.leave_views_module(module)
+
+    def visit_views_module(self, module):
+        """Called when entering a module with a registered view inside it."""
+
+    def leave_views_module(self, module):
+        """Called when leaving from a module with a registered view inside it."""
+
+
+class FunctionBasedViewChecker(DjangoViewsVisiter):
+    __implements__ = (interfaces.IAstroidChecker, )
+
+    name = 'django-function-based-view-checker'
+
+    msgs = {'R4611': ('Function based view used, use class based',
+                      'non-function-based-view-required',
+                      'Function based views are the path to the Dark side!')}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.views_by_module = self.group_views_by_module(self.url_mapping.values())
```

Comment:
Here I am getting wrong results:

```
['tcms.core', 'tcms.kiwi_auth', 'tcms.core.contrib.comments.apps.AppConfig', 'tcms.core.contrib.linkreference', 'tcms.management', 'tcms.testcases.apps.AppConfig', 'tcms.testplans.apps.AppConfig', 'tcms.testruns.apps.AppConfig', 'tcms.telemetry', 'tcms.xmlrpc', 'tcms.telemetry.tests.plugin']

^^^ this is from the left-over print statement above, notice the apps which are installed.

*** views by module
{'tcms.core.admin': {'delete_view', 'add_view'},
 'tcms.core.ajax': {'UpdateTestCaseActorsView',
                    'comment_case_runs',
                    'tags',
                    'update_bugs_to_caseruns'},
 'tcms.core.contrib.comments.views': {'delete', 'post'},
 'tcms.core.contrib.linkreference.views': {'add', 'remove'},
 'tcms.core.views': {'dashboard', 'navigation'},
 'tcms.kiwi_auth.admin': {'delete_view', 'user_change_password'},
 'tcms.kiwi_auth.views': {'LoginViewWithCustomTemplate',
                          'PasswordResetView',
                          'confirm',
                          'profile',
                          'register'},
 'tcms.telemetry.tests.plugin.views': {'example'},
 'tcms.telemetry.views': {'TestingBreakdownView'}}
```

^^^ this is from a print I've added here. Notice how you have both class based and function based views in the list and also how you are missing every app which is declared not by its module name but via its AppConfig.

Reply:
That was my original intention the DjangoViewsVisiter class only gathers views which are directly or transitively included in the project's `urlpatterns`, the idea was that by doing it so you could find views which are not listed in a standard 'views.py' file, e.g. tcms.kiwi_auth.admin's views. If your idea was different, I'll be happy to change the implementation.

If you don't see an module in the self.views_by_module (I'm assuming that this was what you printed out), then the module did not contain views who are routed in the project's urlpatterns.  Can you elaborate on the last AppConfig, are you reffering to [this](https://docs.djangoproject.com/en/2.2/ref/applications/#for-application-users)?

Reply:
I committed a change, that should be able to deal with AppConfig's. Could you please verify the behaviour on the project, I still have issues running it.

Reply:
With the latest update I have the following views by module:

```
{'tcms.core.admin': {'add_view', 'delete_view'},
 'tcms.core.ajax': {'UpdateTestCaseActorsView',
                    'comment_case_runs',
                    'tags',
                    'update_bugs_to_caseruns'},
 'tcms.core.contrib.comments.views': {'delete', 'post'},
 'tcms.core.contrib.linkreference.views': {'add', 'remove'},
 'tcms.core.views': {'navigation', 'dashboard'},
 'tcms.kiwi_auth.admin': {'user_change_password', 'delete_view'},
 'tcms.kiwi_auth.views': {'LoginViewWithCustomTemplate',
                          'PasswordResetView',
                          'confirm',
                          'profile',
                          'register'},
 'tcms.telemetry.tests.plugin.views': {'example'},
 'tcms.telemetry.views': {'TestingBreakdownView'},
 'tcms.testcases.admin': {'add_view', 'change_view'},
 'tcms.testcases.views': {'NewCaseView',
                          'SimpleTestCaseView',
                          'TestCaseExecutionDetailPanelView',
                          'attachment',
                          'clone',
                          'edit',
                          'get',
                          'list_all',
                          'load_more_cases',
                          'printable',
                          'search'},
 'tcms.testplans.admin': {'add_view', 'change_view'},
 'tcms.testplans.views': {'DeleteCasesView',
                          'LinkCasesSearchView',
                          'LinkCasesView',
                          'NewTestPlanView',
                          'ReorderCasesView',
                          'UpdateParentView',
                          'attachment',
                          'clone',
                          'edit',
                          'get',
                          'get_all',
                          'printable',
                          'search'},
 'tcms.testruns.admin': {'add_view', 'change_view', 'delete_view'},
 'tcms.testruns.views': {'AddCasesToRunView',
                         'TestRunReportView',
                         'UpdateAssigneeView',
                         'UpdateCaseRunStatusView',
                         'cc',
                         'change_status',
                         'clone',
                         'edit',
                         'get',
                         'new',
                         'remove_execution',
                         'search',
                         'update_case_run_text'}}
```

Issues:
- `admin.py` is not included directly in urlpatterns and it doesn't define any views so being present is wrong.
- How did `'tcms.telemetry.tests.plugin.views': {'example'},` get included? It is not referenced from `tcms/urls.py` ?